### PR TITLE
Add descriptions of source compatible and binary compatible changes

### LINKED ISSUE #3717: Are there breaking changes / versioning issues in C# 7.2
From @jonskeet via LiveFyre:

It would be really nice to see a section on versioning concerns here. Is changing "public struct" to "readonly public struct" a breaking change? Is adding the "in" modifier to a method parameter a breaking change? I'll investigate for myself, but it would be great to see that information here.

More discussion on the topic:  

Since "in" means the callee is using a reference to the caller's struct rather than a copy of the caller's struct, there is the potential for it to break code.  The key is whether there is any concern for the caller's copy to be changed while the callee is holding a ref to it.  "in" really only says the callee won't/can't change the struct.  It doesn't say the struct can't be changed elsewhere (by the caller or anything else that can access it).  Obviously we're really talking multitasking/multithreading.  In purely single threading nothing else has a chance to be executing while the callee is executing, so nothing else has the opportunity to change the struct.

It feels like more than "potential" to break code - I suspect it's simply not binary compatible, in the same way that changing `Foo(int x)` to `Foo(ref int x)` isn't binary compatible. I suspect this will limit where it can be used for public library projects :(

/cc @VSadov @jaredpar

### PR DESCRIPTION
**Pull request by BillWagner**
Fixes #3717 

Changes in this PR:

**Add article that defines *source breaking change* and *binary breaking change* when using new language features in a library.**  This is ready for a thorough review to make sure I've expressed the concepts in the above issue correctly, and in a manner that is useful.

**Add note about compatibility on some feature descriptions in *what's new* articles.**  This is the second pass of the review, if the concepts are described correctly. I took the following strategy:
1. Only add a note on features that could be part of the public API surface for a type. Other feature descriptions have not been updated. My choice is somewhat subjective. For example, I did not say anything on the feature for new syntax for numeric literals. A numeric literal could be used for a default value of an argument, but that seemed a stretch, and not useful information to add. This PR adds notes when a new feature appears in a member's signature (except for expression bodied members where there have been customer questions).
1. I tried to be concise in calling this out, but still accurate. Question for reviewers: Does that make any of the notes misleading?

The discussion on the issue referenced has examples and details that will be helpful for reviewers.

/cc @jcouv @jaredpar @jonskeet (who wrote the comment for the original issue)

[internal review link - new article](https://review.docs.microsoft.com/en-us/dotnet/csharp/whats-new/version-update-considerations?branch=pr-en-us-7803)
[update - what's new in 6](https://review.docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-6?branch=pr-en-us-7803)
[update - what's new in 7](https://review.docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-7?branch=pr-en-us-7803)
[update - what's new in 7.2](https://review.docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-7-2?branch=pr-en-us-7803)
[update - what's new in 7.3](https://review.docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-7-3?branch=pr-en-us-7803)
What's new in 7.1 was not updated.

### CODE REVIEW

File: docs/csharp/whats-new/version-update-considerations.md
Comment (pkulikov) [inline docs/csharp/whats-new/version-update-considerations.md]: nit: start with "Backwards compatibility..."?

File: docs/csharp/whats-new/version-update-considerations.md
Comment (pkulikov) [inline docs/csharp/whats-new/version-update-considerations.md]: Should we swap two definitions here? First - binary compatible, then - source compatible.
Rationale:
- If it's binary compatible, then it's source compatible. Maybe it's matter of taste; I would start with more strong compatibility.
- "More care may be required when you adopt new language features in a library, and consumers of that library have not yet upgraded" If consumers have not yet upgraded, then source compatibility is a no go. Consumers cannot rebuild, as they have not upgraded. So, better to aim to binary compatibility.
- The **Evaluate your library** section starts with discussion of binary compatibility, then goes on with source compatibility.

File: docs/csharp/whats-new/version-update-considerations.md
Comment (pkulikov) [inline docs/csharp/whats-new/version-update-considerations.md]: Nit: "assemblies ... must be recompiled": is it a wide-spread usage? Or "assembly source code ... must be recompiled to produce a new assembly version"?

File: docs/csharp/whats-new/version-update-considerations.md
Comment (BillWagner) [reply docs/csharp/whats-new/version-update-considerations.md]: > Should we swap two definitions here? First - binary compatible, then - source compatible.

I started writing them in that order. It came out very clumsy, so I switched. I'll try switching again in this round of edits.

**Review by rpetrusha — CHANGES_REQUESTED** [review]
This is a really valuable topic, @BillWagner. (I've never thought about compatibility across compiler versions, only framework versions.) I've left a number of questions, comments, and suggestions.

File: docs/csharp/whats-new/version-update-considerations.md
Comment (BillWagner) [reply docs/csharp/whats-new/version-update-considerations.md]: That's a good idea. I added a those notes in this paragraph.

File: docs/csharp/whats-new/csharp-7-2.md
Comment (BillWagner) [reply docs/csharp/whats-new/csharp-7-2.md]: Added a definition in the compabitility article, and linked here (and other locations where changes on neither source nor binary compatible.

File: docs/csharp/whats-new/version-update-considerations.md
Comment (BillWagner) [reply docs/csharp/whats-new/version-update-considerations.md]: I rewrote the 2nd sentence per @rpetrusha 's comment below. I think that addresses your comment as well. Thoughts?

File: docs/csharp/whats-new/version-update-considerations.md
Comment (BillWagner) [reply docs/csharp/whats-new/version-update-considerations.md]: This comment also helped me re-write the descriptions to reverse the order.  😃

**Comment by BillWagner** [comment]
Thanks @rpetrusha and @pkulikov 

I've udpated with changes that address all the comments.

File: docs/csharp/whats-new/version-update-considerations.md
Comment (pkulikov) [reply docs/csharp/whats-new/version-update-considerations.md]: @BillWagner I like a new edition.

I don't even see the "Resolve conversation" button. I don't remember if it was there before your commit (that might be that the commit has resolved conversation automatically).

Comment (BillWagner) [reply docs/csharp/whats-new/version-update-considerations.md]: Thanks @pkulikov I looks like it's a UI feature that's only shown to the PR author and committers. I've marked this as resolved.

File: docs/csharp/whats-new/version-update-considerations.md
Comment (pkulikov) [inline docs/csharp/whats-new/version-update-considerations.md]: >...you'll need to two variations of compatibility:...

Missing a verb here: "you'll need to consider ..."?

**Review by pkulikov — COMMENTED** [review]
@BillWagner Thank you for addressing comments. I like a new version. Found a couple of nits to consider/fix.

File: docs/csharp/whats-new/version-update-considerations.md
Comment (pkulikov) [reply docs/csharp/whats-new/version-update-considerations.md]: @BillWagner indeed. I see such a button in the PRs created by me.

**Review by rpetrusha — APPROVED** [review]
This is a really good article, @BillWagner. I've left a few more suggested changes, but you can merge when you're ready.

File: docs/csharp/whats-new/version-update-considerations.md
Comment (jcouv) [inline docs/csharp/whats-new/version-update-considerations.md]: Typo: issue

**Review by jcouv — APPROVED** [review]
LGTM modulo typo. Thanks!

**Pull request merged**
